File: app/src/main/kotlin/tech/mmarca/openvitals/features/activity/ActivitySessionMetricChartCards.kt
package tech.mmarca.openvitals.features.activity

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import tech.mmarca.openvitals.R
import tech.mmarca.openvitals.core.presentation.UnitFormatter
import tech.mmarca.openvitals.domain.insights.SplitSource
import tech.mmarca.openvitals.domain.model.ActivityCadenceKind
import tech.mmarca.openvitals.domain.model.ActivityCadenceSample
import tech.mmarca.openvitals.domain.model.SpeedSample
import tech.mmarca.openvitals.ui.components.ChartXAxisWithYAxis
import tech.mmarca.openvitals.ui.components.ChartZoom
import tech.mmarca.openvitals.ui.components.MetricLinePlot
import tech.mmarca.openvitals.ui.components.MetricLinePlotPoint
import tech.mmarca.openvitals.ui.components.OpenVitalsCard
import tech.mmarca.openvitals.ui.components.SessionAxis
import tech.mmarca.openvitals.ui.components.SessionAxisLabels
import tech.mmarca.openvitals.ui.components.SessionPause
import tech.mmarca.openvitals.ui.components.formatElapsedChartLabel
import tech.mmarca.openvitals.ui.theme.CycleColor
import tech.mmarca.openvitals.ui.theme.DistanceColor
import tech.mmarca.openvitals.ui.theme.ElevationColor
import tech.mmarca.openvitals.ui.theme.StepsColor

@Composable
internal fun ActivitySpeedChartCard(
    samples: List<SpeedSample>,
    sessionStart: Instant,
    sessionEnd: Instant,
    unitFormatter: UnitFormatter,
    modifier: Modifier = Modifier,
    pauses: List<SessionPause> = emptyList(),
) {
    if (samples.isEmpty()) return

    ActivitySessionMetricChartCard(
        title = stringResource(R.string.activity_recording_live_speed),
        sortedValues = samples.sortedBy { it.time }.map { it.time to it.metersPerSecond },
        sessionStart = sessionStart,
        sessionEnd = sessionEnd,
        unitFormatter = unitFormatter,
        accentColor = DistanceColor,
        valueFormatter = { unitFormatter.speed(it).text },
        modifier = modifier,
        pauses = pauses,
    )
}

/**
 * Speed rebuilt from the splits for a device that recorded none. The trace
 * steps, one flat run per split, and the stat row counts splits. The
 * trace states its own average: the chart's mean would be wrong.
 */
@Composable
internal fun ActivitySplitSpeedChartCard(
    trace: ActivitySplitSpeedTrace,
    source: SplitSource,
    splitDistanceMeters: Double,
    sessionStart: Instant,
    sessionEnd: Instant,
    unitFormatter: UnitFormatter,
    modifier: Modifier = Modifier,
    pauses: List<SessionPause> = emptyList(),
) {
    if (trace.samples.isEmpty()) return

    // ESTIMATED never reaches here (flat by construction), nor SPEED_SAMPLES (the recorded card wins).
    val title: String
    val countLabel: String
    when (source) {
        SplitSource.DEVICE_LAPS -> {
            title = stringResource(R.string.activity_speed_per_lap_title)
            countLabel = stringResource(R.string.activity_speed_laps_label)
        }
        else -> {
            title = stringResource(
                R.string.activity_speed_per_split_title,
                splitDistanceLabel(unitFormatter, splitDistanceMeters),
            )
            countLabel = stringResource(R.string.activity_speed_splits_label)
        }
    }

    ActivitySessionMetricChartCard(
        title = title,
        sortedValues = trace.samples.map { it.time to it.metersPerSecond },
        sessionStart = sessionStart,
        sessionEnd = sessionEnd,
        unitFormatter = unitFormatter,
        accentColor = DistanceColor,
        valueFormatter = { unitFormatter.speed(it).text },
        countText = unitFormatter.count(trace.splitCount),
        countLabel = countLabel,
        averageOverride = trace.averageMetersPerSecond,
        modifier = modifier,
        pauses = pauses,
    )
}

/**
 * The height profile. Health Connect has no elevation series, so this is
 * drawn from the route's altitudes. No route or altitude, no card.
 */
@Composable
internal fun ActivityElevationChartCard(
    samples: List<ActivityElevationSample>,
    sessionStart: Instant,
    sessionEnd: Instant,
    unitFormatter: UnitFormatter,
    modifier: Modifier = Modifier,
    pauses: List<SessionPause> = emptyList(),
) {
    if (samples.isEmpty()) return

    ActivitySessionMetricChartCard(
        title = stringResource(R.string.metric_elevation),
        sortedValues = samples.map { it.time to it.meters },
        sessionStart = sessionStart,
        sessionEnd = sessionEnd,
        unitFormatter = unitFormatter,
        accentColor = ElevationColor,
        valueFormatter = { unitFormatter.elevation(it).text },
        floorAtZero = false,
        modifier = modifier,
        pauses = pauses,
    )
}

@Composable
internal fun ActivityCadenceChartCard(
    samples: List<ActivityCadenceSample>,
    kind: ActivityCadenceKind,
    sessionStart: Instant,
    sessionEnd: Instant,
    unitFormatter: UnitFormatter,
    modifier: Modifier = Modifier,
    pauses: List<SessionPause> = emptyList(),
) {
    val filtered = samples.filter { it.kind == kind }
    if (filtered.isEmpty()) return

    val title = when (kind) {
        ActivityCadenceKind.CYCLING -> stringResource(R.string.metric_cycling_cadence)
        ActivityCadenceKind.STEPS -> stringResource(R.string.metric_steps_cadence)
    }
    val accentColor = when (kind) {
        ActivityCadenceKind.CYCLING -> CycleColor
        ActivityCadenceKind.STEPS -> StepsColor
    }

    ActivitySessionMetricChartCard(
        title = title,
        sortedValues = filtered.sortedBy { it.time }.map { it.time to it.rate },
        sessionStart = sessionStart,
        sessionEnd = sessionEnd,
        unitFormatter = unitFormatter,
        accentColor = accentColor,
        valueFormatter = { unitFormatter.cadence(it).text },
        modifier = modifier,
        pauses = pauses,
    )
}

@Composable
private fun ActivitySessionMetricChartCard(
    title: String,
    sortedValues: List<Pair<Instant, Double>>,
    sessionStart: Instant,
    sessionEnd: Instant,
    unitFormatter: UnitFormatter,
    accentColor: Color,
    valueFormatter: (Double) -> String,
    modifier: Modifier = Modifier,
    // Speed and cadence cannot be negative. Elevation is read against the
    // ground, not zero, and can be below it.
    floorAtZero: Boolean = true,
    // A stepped trace counts splits, not the corners of the steps.
    countText: String? = null,
    countLabel: String? = null,
    averageOverride: Double? = null,
    pauses: List<SessionPause> = emptyList(),
) {
    if (sortedValues.isEmpty()) return

    val minValue = sortedValues.minOf { it.second }
    val maxValue = sortedValues.maxOf { it.second }
    val avgValue = averageOverride ?: sortedValues.map { it.second }.average()
    val valueRange = (maxValue - minValue).coerceAtLeast(0.001)
    val paddedMin = (minValue - valueRange * 0.1).let { if (floorAtZero) it.coerceAtLeast(0.0) else it }
    val paddedMax = maxValue + valueRange * 0.1
    // The axis counts only moving time: a pause gets none of the chart.
    val axis = remember(sessionStart, sessionEnd, pauses) {
        SessionAxis(start = sessionStart, end = sessionEnd, pauses = pauses)
    }
    val chartHeight = 180.dp
    val zone = ZoneId.systemDefault()
    val timeFormatter = DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT)
    val drawPoints = sortedValues.size <= 120
    // Built once, identity-stable, so the geometry cache holds.
    val chartPoints = remember(sortedValues, axis) {
        sortedValues.map { (time, value) ->
            MetricLinePlotPoint(xFraction = axis.fractionOf(time), value = value)
        }
    }

    OpenVitalsCard(modifier = modifier) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text(
                text = title,
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
            )
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(16.dp),
            ) {
                ActivitySessionMetricStat(
                    label = stringResource(R.string.summary_average),
                    value = valueFormatter(avgValue),
                    accentColor = accentColor,
                    modifier = Modifier.weight(1f),
                )
                ActivitySessionMetricStat(
                    label = stringResource(R.string.summary_range),
                    value = "${valueFormatter(minValue)}-${valueFormatter(maxValue)}",
                    accentColor = accentColor,
                    modifier = Modifier.weight(1f),
                )
                ActivitySessionMetricStat(
                    label = countLabel ?: stringResource(R.string.summary_samples),
                    value = countText ?: unitFormatter.count(sortedValues.size),
                    accentColor = accentColor,
                    modifier = Modifier.weight(1f),
                )
            }
            // Plot and elapsed row share the one viewport.
            ChartZoom(sessionStart, sessionEnd, sortedValues) { zoom ->
                Column {
                    MetricLinePlot(
                        points = chartPoints,
                        minValue = paddedMin,
                        maxValue = paddedMax,
                        accentColor = accentColor,
                        chartHeight = chartHeight,
                        valueFormatter = valueFormatter,
                        pointRadius = if (drawPoints) 2.dp else 0.dp,
                        lineStrokeWidth = 2.dp,
                        drawPoints = drawPoints,
                        viewport = zoom.viewport,
                        multiTouch = zoom.multiTouch,
                        // Moving elapsed, matching the labels under it.
                        scrubLabel = { point ->
                            valueFormatter(point.value) to
                                formatElapsedChartLabel(axis.elapsedAt(point.xFraction))
                        },
                    )
                    Spacer(Modifier.height(4.dp))
                    ChartXAxisWithYAxis {
                        SessionAxisLabels(axis = axis, viewport = zoom.viewport)
                    }
                }
            }
            Text(
                text = stringResource(
                    R.string.summary_recorded,
                    timeFormatter.format(sortedValues.first().first.atZone(zone)),
                    timeFormatter.format(sortedValues.last().first.atZone(zone)),
                ),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

@Composable
private fun ActivitySessionMetricStat(
    label: String,
    value: String,
    accentColor: Color,
    modifier: Modifier = Modifier,
) {
    Column(modifier = modifier) {
        Text(
            text = value,
            style = MaterialTheme.typography.titleMedium,
            color = accentColor,
        )
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

File: lib/features/activity/presentation/activity_session_metric_chart_cards.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

import '../../../core/presentation/unit_formatter.dart';
import '../../../domain/model/activity_models.dart';
import '../../../l10n/app_localizations.dart';
import '../../../ui/charts/chart_axis.dart';
import '../../../ui/charts/metric_line_plot.dart';
import '../../../ui/components/ov_card.dart';
import '../../../ui/theme/app_colors.dart';
import 'activity_heart_rate_chart_card.dart' show sessionElapsedLabels;

/// Port of the Kotlin `ActivitySessionMetricChartCards.kt`: the speed and cadence
/// traces recorded during a session, on the same time axis as the heart-rate card.
///
/// These were dropped in the Flutter port, which is why a cycling activity showed
/// no speed and no cadence even when a sensor had recorded both.

/// Port of the Kotlin `ActivitySpeedChartCard`.
class ActivitySpeedChartCard extends StatelessWidget {
  const ActivitySpeedChartCard({
    super.key,
    required this.samples,
    required this.sessionStart,
    required this.sessionEnd,
    required this.unitFormatter,
  });

  final List<SpeedSample> samples;
  final DateTime sessionStart;
  final DateTime sessionEnd;
  final UnitFormatter unitFormatter;

  @override
  Widget build(BuildContext context) {
    if (samples.isEmpty) return const SizedBox.shrink();
    final sorted = [...samples]..sort((a, b) => a.time.compareTo(b.time));

    return _ActivitySessionMetricChartCard(
      title: AppLocalizations.of(context).activityRecordingLiveSpeed,
      times: [for (final sample in sorted) sample.time],
      values: [for (final sample in sorted) sample.metersPerSecond],
      sessionStart: sessionStart,
      sessionEnd: sessionEnd,
      unitFormatter: unitFormatter,
      accentColor: AppColors.distance,
      valueFormatter: (value) => unitFormatter.speed(value).text,
    );
  }
}

/// Port of the Kotlin `ActivityCadenceChartCard`.
///
/// Takes the whole sample list and filters by [kind] itself, exactly as Kotlin
/// does, so the screen can just render one card per kind. Which kind is right for
/// an activity is not decided by its exercise type anywhere -- Health Connect
/// tags each sample by the record it came from (`CyclingPedalingCadenceRecord` ->
/// cycling, `StepsCadenceRecord` -> steps), so a bike ride simply has no
/// steps-kind samples and the card for it never appears.
class ActivityCadenceChartCard extends StatelessWidget {
  const ActivityCadenceChartCard({
    super.key,
    required this.samples,
    required this.kind,
    required this.sessionStart,
    required this.sessionEnd,
    required this.unitFormatter,
  });

  final List<ActivityCadenceSample> samples;
  final ActivityCadenceKind kind;
  final DateTime sessionStart;
  final DateTime sessionEnd;
  final UnitFormatter unitFormatter;

  @override
  Widget build(BuildContext context) {
    final filtered = [
      for (final sample in samples)
        if (sample.kind == kind) sample,
    ]..sort((a, b) => a.time.compareTo(b.time));
    if (filtered.isEmpty) return const SizedBox.shrink();

    final l10n = AppLocalizations.of(context);
    final (title, accentColor, format) = switch (kind) {
      ActivityCadenceKind.cycling => (
          l10n.metricCyclingCadence,
          AppColors.cycle,
          unitFormatter.cadence,
        ),
      ActivityCadenceKind.steps => (
          l10n.metricStepsCadence,
          AppColors.steps,
          unitFormatter.stepsCadence,
        ),
    };

    return _ActivitySessionMetricChartCard(
      title: title,
      times: [for (final sample in filtered) sample.time],
      values: [for (final sample in filtered) sample.rate],
      sessionStart: sessionStart,
      sessionEnd: sessionEnd,
      unitFormatter: unitFormatter,
      accentColor: accentColor,
      valueFormatter: (value) => format(value).text,
    );
  }
}

/// Port of the Kotlin private `ActivitySessionMetricChartCard`: the renderer both
/// cards share. [times] and [values] are parallel and already sorted by time.
class _ActivitySessionMetricChartCard extends StatelessWidget {
  const _ActivitySessionMetricChartCard({
    required this.title,
    required this.times,
    required this.values,
    required this.sessionStart,
    required this.sessionEnd,
    required this.unitFormatter,
    required this.accentColor,
    required this.valueFormatter,
  });

  final String title;
  final List<DateTime> times;
  final List<double> values;
  final DateTime sessionStart;
  final DateTime sessionEnd;
  final UnitFormatter unitFormatter;
  final Color accentColor;
  final String Function(double value) valueFormatter;

  static const double _chartHeight = 180;

  @override
  Widget build(BuildContext context) {
    if (values.isEmpty) return const SizedBox.shrink();

    final theme = Theme.of(context);
    final l10n = AppLocalizations.of(context);
    final locale = Localizations.localeOf(context).toString();

    final minValue = values.reduce(math.min);
    final maxValue = values.reduce(math.max);
    final avgValue = values.reduce((a, b) => a + b) / values.length;
    // Kotlin pads the Y range by 10% of its span, flooring the minimum at zero.
    // The 0.001 floor keeps a flat series (every sample identical) from
    // collapsing the range to nothing and dividing by zero in the painter.
    final valueRange = math.max(maxValue - minValue, 0.001);
    final paddedMin = math.max(minValue - valueRange * 0.1, 0.0);
    final paddedMax = maxValue + valueRange * 0.1;
    final sessionDurationMillis =
        math.max(sessionEnd.difference(sessionStart).inMilliseconds, 1);
    final drawPoints = values.length <= 120;
    final timeFormatter = DateFormat.jm(locale);

    return OpenVitalsCard(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          spacing: 12,
          children: [
            Text(
              title,
              style: theme.textTheme.titleMedium
                  ?.copyWith(fontWeight: FontWeight.bold),
            ),
            Row(
              spacing: 16,
              children: [
                Expanded(
                  child: _ActivitySessionMetricStat(
                    label: l10n.summaryAverage,
                    value: valueFormatter(avgValue),
                    accentColor: accentColor,
                  ),
                ),
                Expanded(
                  child: _ActivitySessionMetricStat(
                    label: l10n.summaryRange,
                    value: '${valueFormatter(minValue)}-'
                        '${valueFormatter(maxValue)}',
                    accentColor: accentColor,
                  ),
                ),
                Expanded(
                  child: _ActivitySessionMetricStat(
                    label: l10n.summarySamples,
                    value: unitFormatter.count(values.length),
                    accentColor: accentColor,
                  ),
                ),
              ],
            ),
            MetricLinePlot(
              points: [
                for (var i = 0; i < values.length; i++)
                  MetricLinePlotPoint(
                    xFraction: times[i]
                            .difference(sessionStart)
                            .inMilliseconds
                            .clamp(0, sessionDurationMillis) /
                        sessionDurationMillis,
                    value: values[i],
                  ),
              ],
              minValue: paddedMin,
              maxValue: paddedMax,
              accentColor: accentColor,
              chartHeight: _chartHeight,
              valueFormatter: valueFormatter,
              pointRadius: drawPoints ? 2 : 0,
              lineStrokeWidth: 2,
              drawPoints: drawPoints,
            ),
            const SizedBox(height: 4),
            ChartXAxisWithYAxis(
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  for (final label in sessionElapsedLabels(
                      Duration(milliseconds: sessionDurationMillis)))
                    Text(
                      label,
                      style: theme.textTheme.labelSmall?.copyWith(
                        color: theme.colorScheme.onSurfaceVariant,
                      ),
                    ),
                ],
              ),
            ),
            Text(
              l10n.summaryRecorded(
                timeFormatter.format(times.first.toLocal()),
                timeFormatter.format(times.last.toLocal()),
              ),
              style: theme.textTheme.bodySmall?.copyWith(
                color: theme.colorScheme.onSurfaceVariant,
              ),
            ),
          ],
        ),
      ),
    );
  }
}

/// Kotlin `ActivitySessionMetricStat`.
class _ActivitySessionMetricStat extends StatelessWidget {
  const _ActivitySessionMetricStat({
    required this.label,
    required this.value,
    required this.accentColor,
  });

  final String label;
  final String value;
  final Color accentColor;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          value,
          style: theme.textTheme.titleMedium?.copyWith(color: accentColor),
        ),
        Text(
          label,
          style: theme.textTheme.labelSmall?.copyWith(
            color: theme.colorScheme.onSurfaceVariant,
          ),
        ),
      ],
    );
  }
}
